src/k4a_ros_device.cpp

```diff
@@ -799,6 +803,53 @@ k4a_result_t K4AROSDevice::getBodyMarker(const k4abt_body_t& body, MarkerPtr mar
 
     return K4A_RESULT_SUCCEEDED;
 }
+
+k4a_result_t K4AROSDevice::getBodyIndexMap(const k4abt::frame& body_frame, sensor_msgs::ImagePtr body_index_map_image)
+{
+    k4a::image k4a_body_index_map = body_frame.get_body_index_map();
+
+    if (!k4a_body_index_map)
+    {
+        ROS_ERROR("Cannot render body index map: no body index map");
+        return K4A_RESULT_FAILED;
+    }
+
+    return renderBodyIndexMapToROS(body_index_map_image, k4a_body_index_map, body_frame);
+}
+
+k4a_result_t K4AROSDevice::renderBodyIndexMapToROS(sensor_msgs::ImagePtr body_index_map_image, k4a::image& k4a_body_index_map, const k4abt::frame& body_frame)
+{
+    // Access the ir image as an array of uint16 pixels
+    BodyIndexMapPixel* body_index_map_frame_buffer = k4a_body_index_map.get_buffer();
+    auto body_index_map_pixel_count = k4a_body_index_map.get_size() / sizeof(BodyIndexMapPixel);
+
+    // Build the ROS message
+    body_index_map_image->height = k4a_body_index_map.get_height_pixels();
+    body_index_map_image->width = k4a_body_index_map.get_width_pixels();
+    body_index_map_image->encoding = sensor_msgs::image_encodings::MONO8;
+    body_index_map_image->is_bigendian = false;
+    body_index_map_image->step = k4a_body_index_map.get_width_pixels() * sizeof(BodyIndexMapPixel);
+
+    // Enlarge the data buffer in the ROS message to hold the frame
+    body_index_map_image->data.resize(body_index_map_image->height * body_index_map_image->step);
+
+    // TODO: can this be done faster?
+    for (size_t i = 0; i < body_index_map_pixel_count; ++i)
+    {
+        BodyIndexMapPixel val = body_index_map_frame_buffer[i];
+        if (val == K4ABT_BODY_INDEX_MAP_BACKGROUND)
+        {
+            body_index_map_image->data[i] = K4ABT_BODY_INDEX_MAP_BACKGROUND;
+        }
+        else
+        {
+            auto body_id = k4abt_frame_get_body_id(body_frame.handle(), val);
+            body_index_map_image->data[i] = body_id % K4ABT_BODY_INDEX_MAP_BACKGROUND;
```

Comment:
Using `body_id % K4ABT_BODY_INDEX_MAP_BACKGROUND` to fit the `body_id` into a `uint8_t` will result in collisions between some body ID's (ex: 1, 256). 

How often do you expect to see body ID's greater than 255? If it's frequent, can we increase the data-width of the image (say, to MONO16) so that we can render more ID's before seeing a collision?

Should the node emit a warning if there's a body-index collision?

Reply:
Good question. I am not sure about this either. If we would use just the index and not the body id we would not have the problem but then we can't relate to the body joint data. But I don't know how important that is.

So we could either:
- Change it to mono16 to enlargen the space for IDs
- Change it to use body index instead of ID but the receiving nodes don't have the body tracking SDK instance to retrieve the corresponding Body ID then

Reply:
Fair enough... lets stick with the % 255 for now and see how well it works in practice. If we encounter issues down the road, we can open a bug and make changes.